qiskit/circuit/quantumcircuitdata.py

```diff
@@ -26,6 +26,27 @@ class CircuitInstruction:
     """A single instruction in a :class:`.QuantumCircuit`, comprised of the :attr:`operation` and
     various operands.
 
+    .. note::
+
+        There is some possible confusion in the names of this class, :class:`~.circuit.Instruction`,
+        and :class:`~.circuit.Operation`, and this class's attribute :attr:`operation`.  Our
+        preferred terminology is by analogy to assembly languages, where an "instruction" is made up
+        of an "operation" and its "operands".
+
+        Historically, :class:`~.circuit.Instruction` came first, and originally contained the qubits
+        it operated on and any parameters, so it was a true "instruction".  Over time,
+        :class:`.QuantumCircuit` became responsible for tracking qubits and clbits, and the class
+        became better described as an "operation".  Changing the name of such a core object would be
+        a very unpleasant API break for users, and so we have stuck with it.
+
+        This class was created to provide a formal "instruction" context object in
+        :class:`.QuantumCircuit.data`, which had long been made of ad-hoc tuples.  With this, and
+        the advent of the :class:`~.circuit.Operation` interface for adding more complex objects to
+        circuits, we took the opportunity to correct the historical naming.  For the time being,
+        this leads to an awkward case where :attr:`.CircuitInstruction.operation` is often an
```

Comment:
You say `operation` is "often" an `Instruction` instance, but its type annotation indicates that it is _always_ an `Instruction` instance.

Reply:
Ah, thanks - that's just a mistake because #7966 (formally adding `Operation`) ended up merging after the PR that added this (#8093) and it didn't get updated.  I'll fix that in a separate PR - it looks like there's also a couple of places in the legacy handling of the `QuantumCircuitData` wrapper that didn't get updated with #7966 either and I can squash them all in one go.